### ISSUE TITLE
code-gen P4.3: publish composite-preference datasets to HuggingFace Hub

### ISSUE THREAD

**Issue by gotoplanb** [gotoplanb/conduct#41]
**Part of #22** · Tags: \`[Cognitive]\` · Builds on #31

The current export contract (\`GET /datasets/preferences\` per [\`docs/datasets.md\`](https://github.com/gotoplanb/conduct/blob/main/docs/datasets.md)) is admin-only — appropriate for one-off pulls, but the wrong primitive for the **cognitive flywheel**. Multi-tenant clients (like [wander](https://github.com/gotoplanb/wander)) shouldn't need admin credentials to consume their own training data, and exposing admin routes to external networks broadens the trust boundary unnecessarily.

## Ask

Conduct **publishes** preference exports to a configured HuggingFace dataset repo on a schedule it controls, using its **own** HF token. Clients pull from HF Hub like any other public dataset — no Conduct creds in the loop.

Conceptual flow:

\`\`\`
Conduct                                  HuggingFace Hub                Wander
───────                                  ───────────────                ──────
[scheduled]                            datasets/gotoplanb/             (huggingface_hub
GET /datasets/preferences  ──►         conduct-code-gen-prefs-v1   ◄──  load_dataset)
push to HF (Conduct's HF token)        @<sha-or-version-tag>            no Conduct creds
\`\`\`

## Config sketch (\`config/seed.routing.yaml\` or a new \`config/exports.yaml\`)

\`\`\`yaml
exports:
  - kind: preferences
    task_type: code_generation
    method: composite
    schedule: \"@daily\"          # or cron
    destination:
      type: huggingface
      repo: gotoplanb/conduct-code-gen-prefs   # configurable per export
      private: true                              # toggle
      tag_pattern: \"v{{date}}-{{count}}rows\"
    filters:
      min_gap: 2
      include_provenance: true
\`\`\`

## Acceptance

- A scheduled job exports preferences and pushes them to the configured HF repo as a versioned snapshot (commit + tag).
- The HF dataset card surfaces provenance back to source job_ids (or hashes of them).
- HF token is supplied via env (\`HF_TOKEN\`); not stored in the routing config.
- A clear error log entry when push fails (HF rate-limit, network, auth) — doesn't drop the rows.
- One Conduct-side test that the export contents match what \`GET /datasets/preferences\` would return for the same params.

## Why this is the right primitive

- **No cross-tenant trust.** Wander has its own HF identity, doesn't see Conduct internals, can't accidentally pull other tenants' data.
- **No admin routes exposed to the world.** \`/datasets/preferences\` can stay admin-only for ops use; the cognitive flywheel uses the public HF path.
- **Versioning is free.** HF Hub already does revision pinning. Reproducibility for fine-tune runs comes from \`load_dataset(\"repo\", revision=\"<sha>\")\`.
- **Cleaner deletes.** If a row needs to be retracted (PII, mistake), Conduct republishes; old HF revision is preserved for audit but no longer the default.
- **One pattern for both directions.** Pairs naturally with the inverse: register HF-hosted fine-tunes as routing targets (separate issue, supersedes #32).

## Related

- [conduct#31](https://github.com/gotoplanb/conduct/issues/31) — the composite-preference shape this exports (closed)
- [wander#7](https://github.com/gotoplanb/wander/issues/7) — the DPO consumer
- [wander/bench/dpo/](https://github.com/gotoplanb/wander/tree/main/bench/dpo) — current scaffold that would migrate to \`load_dataset(...)\` instead of admin-curl

🤖 Filed via [Claude Code](https://claude.com/claude-code)

**Comment by gotoplanb** [comment]
## Addressed — 3f75839 — but via client-key export, not HuggingFace

Dave's steer: the real concern isn't "we need HF," it's "Wander shouldn't need an **admin token** to get its data — that breaks the tenancy wall; services should talk via **client keys**." HF was a workaround to avoid the admin route. The Conduct-native fix is simpler and is the one shipped:

**`/datasets/sft` + `/datasets/preferences` now authenticate with `current_client_or_admin`:**
- A **client key** pulls *its own* jobs' data only (scoped by `client_app_id`) — no admin token crosses the boundary, no cross-tenant leakage. This is the SaaS path.
- The **admin key** stays unscoped (all tenants) for ops.

So Wander consumes its training data with its normal client credential: `GET /datasets/preferences?method=composite&task_type=code_generation` + `Authorization: Bearer <wander-client-key>`. Works across hosts (it's just HTTP + a client key — no shared filesystem, no admin creds, no SSH).

**Live-verified:** no-auth → 403; admin → 74 rows (unscoped); a fresh client key → 200 with 0 rows (its own data); cross-tenant isolation covered by tests.

### On HuggingFace specifically
Not built — and per Dave it's a *suggestion*, not a requirement. A client-key pull is the right primitive whether the consumer is local or remote, and it avoids adding a cloud dependency + token + scheduled-publish machinery to a local-first system. **If** you later want HF for its *other* value — public/versioned/shareable dataset artifacts (e.g. publishing the flywheel's outputs as research) — that's a clean separate issue with that as the explicit goal, not as an auth workaround. Closing this one: the stated problem (data access without admin creds) is solved.
